keps/sig-network/563-dual-stack/README.md

```diff
@@ -370,16 +390,39 @@ The PodStatus internal representation will be modified to use a slice of
 PodIP structs rather than a singular IP ("PodIP"):
 
 ```
-    // IP address information. Each entry includes:
-    //    IP: An IP address allocated to the pod. Routable at least within
-    //        the cluster.
-    // Empty if no IPs have been allocated yet.
-    type PodIP struct {
-        IP string
-    }
+// IP address information for entries in the (plural) PodIPs field.
+// Each entry includes:
+//    IP: An IP address allocated to the pod. Routable at least within the cluster.
+type PodIP struct {
+	// ip is an IP address (IPv4 or IPv6) assigned to the pod
+	IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`
+}
+
+// podIPs holds the IP addresses allocated to the pod. If this field is specified, the 0th entry must
+// match the podIP field. Pods may be allocated at most 1 value for each of IPv4 and IPv6. This list
+// is empty if no IPs have been allocated yet.
+// +optional
+// +patchStrategy=merge
+// +patchMergeKey=ip
+PodIPs []PodIP `json:"podIPs,omitempty" protobuf:"bytes,12,rep,name=podIPs" patchStrategy:"merge" patchMergeKey:"ip"`
+
+// IP address information for entries in the (plural) HostIPs field.
+// Each entry includes:
+//    IP: An IP address allocated to the host.
+type HostIP struct {
+	// ip is an IP address (IPv4 or IPv6) assigned to the host
+	IP string `json:"ip,omitempty" protobuf:"bytes,1,opt,name=ip"`
+}
+
+
+// hostIPs holds the IP addresses allocated to the host. If this field is specified, the 0th entry must
+// match the hostIP field. Host may be allocated at most 1 value for each of IPv4 and IPv6. This list
```

Comment:
> Host may be allocated at most 1 value for each of IPv4 and IPv6

That could be problematic: nearly all IPv6-enabled nodes will have multiple IPv6 addresses (at least one global unicast address, plus a link-local address per interface that has IPv6 enabled).

Reply:
I think he refers to the NodeIP, the one used by kubelet, it always pick one from each family

Reply:
My worry overall is really about strengthening an assumption that doesn't really hold. [RFC 7934](https://tools.ietf.org/html/rfc7934) recommends that “_networks provide general-purpose end hosts with multiple global IPv6 addresses when they attach_”. One of several mechanisms for this is for an IPv6-enabled host to acquire one IPv6 address using SLAAC and another via DHCPv6, both being valid.

IPv6 is also great for letting you run servers that are trivially dual-homed: set up two external network interfaces on different networks, and autoconfigure them. In a kubernetes cluster you might, separately to node configuration, make have the API server(s) also reachable via separately addressed IPv6 networks, providing resilience even if one networking subsystem fails.

Nodes _aren't_ general-purpose; however, if we're narrowing from the IETF recommendations let's explicitly make that decision and document it.

https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/  doesn't provide a good mechanism for declaring the IPv4 and IPv6 addresses for a kubelet, so my assumption is that the kubelet on a dual-stack node has to pick an address based on a heuristic. I'm also inferring that administrators won't have a good mechanism to avoid that heuristic; much less to declare that both (or more) IPv6 addresses are actually valid.

If we declare that there is always exactly one node IP address per address family, and make it part of the Kubernetes API, we have to live with that decision. I'm calling out that out.

Reply:
> If we declare that there is always exactly one node IP address per address family, and make it part of the Kubernetes API, we have to live with that decision. I'm calling out that out.

right now kubelet takes one node IP address per IP family and I think that is the right thing to do, it simplifies operation and management of the cluster .... I'm coming from large routing networks, routers can have hundreds of IPs, best practice is to assign the "main/control-plane" IP of the router to the loopback interface, so you have consistency in your monitoring/scripts/... and this IP is always reachable through any of the other IPs in the node, that is the admin duty, to handle these other IPs ... 
I think you are right if we consider that the node API should represent a node as per RFC, but I digress here, it should represent a Kubernetes Node, all the other IPs are part of the host not of the "Kubernetes Node", they may change in the future, be down or up, the important is the NodeIP

Reply:
In that case, do we want to change the kubelet command line options / config to allow specifying which address to use as the NodeIP, per address family?

(this might already be done but without documentation)

Reply:
Also, thinking about it, considerations for _routers_ don't translate well to be considerations for nodes. Nodes can act as a router but I think of them much more like hosts.

Reply:
> Also, thinking about it, considerations for routers don't translate well to be considerations for nodes. Nodes can act as a router but I think of them much more like hosts.

ALL kubernetes nodes are routers :) pods have interfaces too , actually ip_forwarding true is a requirement

> In that case, do we want to change the kubelet command line options / config to allow specifying which address to use as the NodeIP, per address family

https://github.com/kubernetes/kubernetes/pull/95239

Reply:
I'm not sure, but should we also update [`kubelet.config.k8s.io`](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/) to allow setting dual-stack values for `address`?

(again, maybe this is implemented but not well documented).

Reply:
yeah, it is not easy :) , see the long discussion and the outcome
https://github.com/kubernetes/kubernetes/pull/95768

> I'm not sure, but should we also update kubelet.config.k8s.io to allow setting dual-stack values for address?
address is not equivalent to the node-ip flag is it?
I think that is more similar to the apiserver problem https://github.com/kubernetes/enhancements/pull/2622

(sorry for throwing you hundreds of links, you are raising totally valid points that are/were discussed other threads, trying to keep the discussions on their own thread)

Reply:
> If we declare that there is always exactly one node IP address per address family, and make it part of the Kubernetes API, we have to live with that decision. I'm calling out that out.

We do not declare that; we declare that there is always exactly one _canonical_ node IP address per family, which (unless you've overridden the default behavior with other config options) can be used to reach the kubelet, hostNetwork pods, pod HostPorts, NodePort services, and a few other things. `Node.Status.Addresses` can also list an arbitrary number of additional node IPs, which we make no claims about, and nodes can also have other IPs not explicitly known to kubernetes.

Basically, `hostIPs` says "if you want to contact the host this pod is on, these IPs will work". It does not say anything about whether other IPs will also work. There was a discussion somewhere recently about whether it should also imply "if the host contacts you, it will use one of these source IPs" but I don't remember where that was...

Reply:
Ah, that's a really good clarification. Thanks @danwinship.